Breadcrumbs

LotteHofstede opened the pull request:
This PR adds breadcrumbs to DSpace. It depends on #578 and https://github.com/DSpace/DSpace/pull/2673

There are two types of breadcrumbs at the moment: 

- i18n breadcrumbs
	- For i18n breadcrumbs the key prefixed with '`breadcrumb.'` and resolved to a message
	- To create this type of breadcrumb, add the following to the router module of the page 
	  `resolve: {breadcrumb: I18nBreadcrumbResolver}, data: {breadcrumbKey: 'i18n.string.to.use.as.key'}` 
- DSO breadcrumbs
	- For DSO's a breadcrumb is added for each parent object and the object self 
	- To create this type of breadcrumb, add the following to the router module of i.e. an item page
    `resolve: {breadcrumb: ItemBreadcrumbResolver}` 

TODOs
- Write tests
- Add TypeDocs
- Add translations for the i18n keys
- Style for mantis

tdonohue commented:
I know this is still a Draft, but just wanted to note that I just merged #578 , so this can be rebased to make it a bit smaller / easier to start reviewing.

tdonohue commented on src/app/breadcrumbs/breadcrumb/breadcrumb-config.model.ts:
```diff
@@ -0,0 +1,7 @@
+import { BreadcrumbsService } from '../../core/breadcrumbs/breadcrumbs.service';
+
+export interface BreadcrumbConfig<T> {
```
Please add TypeDocs for this new interface

tdonohue commented on src/app/breadcrumbs/breadcrumb/breadcrumb.model.ts:
```diff
@@ -0,0 +1,6 @@
+export class Breadcrumb {
```
Please add TypeDocs for this new class

tdonohue commented on src/app/core/shared/child-hal-resource.model.ts:
```diff
@@ -0,0 +1,5 @@
+import { HALResource } from './hal-resource.model';
+
+export interface ChildHALResource extends HALResource {
```
Please add TypeDocs to this new interface

LotteHofstede commented:
Thanks @tdonohue for your feedback!
I added some missing TypeDocs you requested.

I'll address your issues in a similar list:

1. I've now implemented this similarly as to what you requested. The breadcrumbs are now hidden if their path is not complete (meaning there is no breadcrumb specified for the last part of the route). You can still force incomplete paths to show by setting 'showBreadcrumbs: true' on the data. 

2. Browse by: This was a tricky one. Unfortunately the implementation of the browse-by pages for DSO's is not optimal, in my opinion. At the moment, the component cannot differentiate between communities and collection. This also means there is no way of showing different metadata for different types of DSOs on browse pages. 
To solve the issue for my use case, I added separate breadcrumb resolvers to create a logical trail. I do think that this could have been a lot easier if we refactored the DSO browse-by pages a bit. 
 
3. The delay has to do with the breadcrumbs being loaded asynchronously. We chose to implement it like this, so the page does not have to wait for all DSO's to be resolved before it can load. We could change this, but that might cause a lot of pages to load a lot slower.

4. I didn't style the current breadcrumbs other than using the basic bootstrap classes for breadcrumbs, in order to keep it as clean as possible. I'm planning on adding a better style for the mantis theme.
